backend/src/git.rs

```diff
@@ -332,65 +408,233 @@ impl Interface {
     fn git_pull(repo: &Repository) -> Result<()> {
         // https://github.com/rust-lang/git2-rs/blob/master/examples/pull.rs
         // TODO: configure branch via environment variables
-        let fetch_head = Self::git_fetch(repo)?;
+        let fetch_head = Self::git_fetch(repo, None)?;
         info!("Successfully fetched latest changes, merging...");
         Self::git_merge(repo, "master", fetch_head)?;
         info!("Successfully merged latest changes");
         Ok(())
     }
 
-    /// A code level re-implementation of `git fetch`. `git fetch` will sync your local `origin/[BRANCH]` with the remote, but it won't
-    /// merge those changes into `main`.
+   /// Pull the latest changes for a specified branch in the repository.
+    ///
+    /// This function performs a series of operations that mimic the behavior of the `git pull` command:
+    /// 1. **Reset Local Changes**: Discards any local changes in the working directory to ensure that the 
+    ///    repository is in a clean state before pulling new changes.
+    /// 2. **Check Branch Existence**: Verifies whether the specified local branch exists. If it does not,
+    ///    an error is returned.
+    /// 3. **Set Upstream Tracking**: Attempts to set the upstream tracking reference for the specified branch 
+    ///    if it is not already set. This allows the local branch to track changes from the corresponding 
+    ///    remote branch.
+    /// 4. **Fetch Changes**: Retrieves the latest changes from the remote repository for the specified branch.
+    /// 5. **Reset Local Branch**: Resets the local branch to match the state of the upstream branch, effectively 
+    ///    discarding any local commits that are not present in the upstream branch.
     ///
-    /// This implementation fetches all branches.
+    /// # Parameters
+    /// - `branch`: A string slice that represents the name of the local branch to pull changes for.
     ///
-    /// Returns a reference to the latest commit fetched from remote (`FETCH_HEAD`). This is done if you'd like to merge the remote changes into a local branch.
-    fn git_fetch(repo: &Repository) -> Result<AnnotatedCommit<'_>> {
+    /// # Errors
+    /// This function will return an error if any of the following occur:
+    /// - The specified branch does not exist.
+    /// - There are issues with resetting the repository or finding references.
+    /// - Fetching changes from the remote repository fails.
+    #[tracing::instrument(skip(self))]
+    pub fn git_pull_branch(&self, branch: &str) -> Result<()> {
+        // Lock and check the repository
+        let repo = self.repo.lock().unwrap();
+        
+        debug!("Current repository state: {:?}", repo.state()); //leaving here incase of errors
```

Comment:
IMO the comment here could be removed
```suggestion
        debug!("Current repository state: {:?}", repo.state());
```